File: 3583-count-special-triplets/3583-count-special-triplets.cpp
using ll = long long;
const ll MOD = 1e9 + 7;
class Solution {
public:
    int specialTriplets(vector<int>& nums) {
        int n = nums.size();
        vector<ll> ltable(200001, 0), rtable(200001, 0);
        for(int num : nums) {
            rtable[num]++;
        }

        ll ans = 0;
        rtable[nums[0]]--;
        ltable[nums[0]]++;
        for(int j = 1; j < n-1; j++) {
            rtable[nums[j]]--;
            ans = (ans + ltable[nums[j]*2] * rtable[nums[j]*2]) % MOD;
            ltable[nums[j]]++;
        }

        return ans;
    }
};

File: 3583-count-special-triplets/3583-count-special-triplets.go
const MOD int = 1e9 + 7

func specialTriplets(nums []int) int {
    n := len(nums)
    lTable := make([]int, 200001) 
    rTable := make([]int, 200001) 

    for _, num := range nums {
        rTable[num]++
    }
    rTable[nums[0]]--
    lTable[nums[0]]++

    ans := 0
    for j := 1; j < n-1; j++ {
        rTable[nums[j]]--
        target := nums[j] * 2
        ans = (ans + lTable[target] * rTable[target]) % MOD
        lTable[nums[j]]++
    } 
    return ans
}